src/pure_aworset.erl

```diff
@@ -0,0 +1,219 @@
+%% Copyright (c) 2007-2012 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc Pure AWORSet CRDT: pure op-based add-wins observed-remove set
+%%
+%% @reference Carlos Baquero, Paulo Sérgio Almeida, and Ali Shoker
+%%      Making Operation-based CRDTs Operation-based (2014)
+%%      [http://haslab.uminho.pt/ashoker/files/opbaseddais14.pdf]
+
+-module(pure_aworset).
+-author("Georges Younes <georges.r.younes@gmail.com>").
+
+-behaviour(type).
+-behaviour(pure_polog).
+
+-define(TYPE, ?MODULE).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([new/0, new/1]).
+-export([mutate/3, query/1, equal/2]).
+-export([redundant/2, remove_redundant_crystal/2, remove_redundant_polog/2, check_stability/2]).
+
+-export_type([pure_aworset/0, pure_aworset_op/0]).
+
+-opaque pure_aworset() :: {?TYPE, payload()}.
+-type payload() :: {pure_type:polog(), ordsets:set()}.
+-type pure_aworset_op() :: {add, pure_type:element()} | {rmv, pure_type:element()}.
+
+%% @doc Create a new, empty `pure_aworset()'
+-spec new() -> pure_aworset().
+new() ->
+    {?TYPE, {orddict:new(), ordsets:new()}}.
+
+%% @doc Create a new, empty `pure_aworset()'
+-spec new([term()]) -> pure_aworset().
+new([]) ->
+    new().
+
+%% @doc Check redundancy `pure_aworset()'
+%% Called in remove_redundant().
+-spec redundant({pure_type:id(), pure_aworset_op()}, {pure_type:id(), pure_aworset_op()}) ->
+    integer().
+redundant({VV1, {add, Elem1}}, {VV2, {_X, Elem2}}) ->
+    case Elem1 == Elem2 andalso pure_trcb:happened_before(VV1, VV2) of
+        true ->
+            1;
+        false ->
+            0
+    end.
+
+%% @doc Removes redundant operations from POLog of `pure_aworset()'
+%% Called upon updating (add, rmv) the `pure_aworset()'
+-spec remove_redundant_polog({pure_type:id(), pure_aworset_op()}, pure_aworset()) -> {boolean(), pure_aworset()}.
+remove_redundant_polog({VV1, Op}, {?TYPE, {POLog0, ORSet}}) ->
+    POLog1 = orddict:fold(
+        fun(Key, Value, Acc) ->
+            case redundant({Key, Value}, {VV1, Op}) of
+                0 ->
+                    orddict:store(Key, Value, Acc);
+                1 ->
+                    Acc
+            end
+        end,
+        orddict:new(),
+        POLog0
+    ),
+    {true, {?TYPE, {POLog1, ORSet}}}.
```

Comment:
What does the return value of `true` indicate?  If it's always true, does it need to be there?